Implement request methods content blocker trigger field

elijahsawyers opened the pull request:
#### 301771c71cc7b6b56aa1e2567926fc71d927d459
<pre>
Implement request methods content blocker trigger field
<a href="https://bugs.webkit.org/show_bug.cgi?id=290962">https://bugs.webkit.org/show_bug.cgi?id=290962</a>
<a href="https://rdar.apple.com/148474901">rdar://148474901</a>

Reviewed by Alex Christensen.

This patch introduces a new trigger field, request-method, to perform an action
only if the HTTP request method matches the specified HTTP request method. This
field has a string value, which can be any of the HTTP methods (lowercased) as
defined by:

<a href="https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Methods">https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Methods</a>

We chose to use a string value, rather than an array of strings, because making
the request methods mutually exclusive saves more space (4 bits instead of 9).
Although this pattern deviates from the trigger field norm, it&apos;s better to save
the bits since we&apos;re running out of room.

The motivation behind this change is that we need to support requestMethods and
excludedRequestMethods in dNR, which isn&apos;t possible without this change.

* Source/WebCore/contentextensions/ContentExtensionError.cpp:
(WebCore::ContentExtensions::contentExtensionErrorCategory):
* Source/WebCore/contentextensions/ContentExtensionError.h:
Introduce a new error for invalid request-method field.

* Source/WebCore/contentextensions/ContentExtensionParser.cpp:
(WebCore::ContentExtensions::loadTrigger):
Parse the request-method into the trigger flags, raising an error as needed.

* Source/WebCore/contentextensions/ContentExtensionsBackend.cpp:
(WebCore::ContentExtensions::ContentExtensionsBackend::processContentRuleListsForLoad):
(WebCore::ContentExtensions::ContentExtensionsBackend::processContentRuleListsForPingLoad):
Parse the HTTP request method of the load and include it in the resource load
info.

* Source/WebCore/contentextensions/ContentExtensionsBackend.h:
Make the method for processing content rule lists for ping loads take in an HTTP
method parameter.

* Source/WebCore/contentextensions/DFABytecode.h:
* Source/WebCore/contentextensions/DFABytecodeCompiler.cpp:
(WebCore::ContentExtensions::bytecodeFlagsSize):
(WebCore::ContentExtensions::toSizeT):
* Source/WebCore/contentextensions/DFABytecodeInterpreter.cpp:
(WebCore::ContentExtensions::consumeResourceFlagsAndInstruction):
(WebCore::ContentExtensions::DFABytecodeInterpreter::interpretTestFlagsAndAppendAction):
Introduce a 32 bit flags size. With the introduction of request methods, the
flags can be up to 25 bits in size.

* Source/WebCore/loader/ResourceLoadInfo.cpp:
(WebCore::ContentExtensions::readRequestMethod):
(WebCore::ContentExtensions::ResourceLoadInfo::getResourceFlags const):
* Source/WebCore/loader/ResourceLoadInfo.h:
(WebCore::ContentExtensions::ResourceLoadInfo::isolatedCopy const):
(WebCore::ContentExtensions::ResourceLoadInfo::isolatedCopy):
Introduce the request methods enum, and modify the byte code as necessary.

* Source/WebKit/NetworkProcess/NetworkLoadChecker.cpp:
(WebKit::NetworkLoadChecker::processContentRuleListsForLoad):
Pass in the HTTP method when processing content rule lists for ping load.

* Tools/TestWebKitAPI/Tests/WebCore/ContentExtensions.cpp:
(TestWebKitAPI::TEST_F(ContentExtensionTest, RequestMethod)):
Write a new test for validating the behavior of the request-method field.

(TestWebKitAPI::TEST_F(ContentExtensionTest, InvalidJSON)):
Add assertions for validating the request-method field in the JSON.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/ContentRuleListNotification.mm:
(TEST(ContentRuleList, RequestMethods)):
Write a new test for validating the behavior of the request-method field.

Canonical link: <a href="https://commits.webkit.org/293746@main">https://commits.webkit.org/293746@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f313f23d9b7981162cc8e0ea302257006a99ecf5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99774 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19419 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9731 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104899 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50358 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101815 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19724 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27855 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75947 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33046 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102781 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15024 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90124 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56312 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14832 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8111 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49724 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84758 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8196 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107260 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26885 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19632 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84903 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27248 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86329 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84428 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21442 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29102 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/6840 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/20694 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26824 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32034 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26635 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29952 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28200 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

elijahsawyers commented on Source/WebKit/UIProcess/API/APIContentRuleListStore.h:
```diff
@@ -54,7 +54,7 @@ class ContentRuleListStore final : public ObjectImpl<Object::Type::ContentRuleLi
     // This should be incremented every time a functional change is made to the bytecode, file format, etc.
     // to prevent crashing while loading old data.
-    static constexpr uint32_t CurrentContentRuleListFileVersion = 19;
+    static constexpr uint32_t CurrentContentRuleListFileVersion = 20;
```
We might not need this. I didn't shuffle any of the bits around, so I think these changes should be backwards compatible, no?

xeenon replied:
Yeah we should be okay to not bump it, since we will read `0`s for the request method part (which means no method condition).

elijahsawyers commented on Source/WebCore/contentextensions/ContentExtensionsBackend.cpp:
```diff
@@ -235,6 +235,7 @@ ContentRuleListResults ContentExtensionsBackend::processContentRuleListsForLoad(
     URL mainDocumentURL;
     URL frameURL;
     bool mainFrameContext = false;
+    OptionSet<RequestMethod> requestMethod = readRequestMethod(initiatingDocumentLoader.request().httpMethod().convertToASCIILowercase()).value();
```
Is it ever possible that `initiatingDocumentLoader.request().httpMethod()` would be `null`?

xeenon commented on Source/WebCore/loader/ResourceLoadInfo.h:
```diff
@@ -72,19 +72,33 @@ enum class LoadContext : uint32_t {
+    Patch = 0x10000000,
+    Connect = 0x20000000,
+};
+static constexpr uint32_t RequestMethodMask = 0x3FE00000;
```
These are mutually exclusive, so we can use a much smaller footprint — only 4 bits. But even that might be more than we need, given that most of these methods are rarely used. It could make sense to collapse this into a simpler set — `Get`, `Post`, `Connect`, and maybe an `Other` — which would only require 2 bits. We would just need to consider the risks vs rewards to dNR of not supporting the other methods.

Either way, much better than 9 bits as you have it.

elijahsawyers replied:
It looks like Chrome defines: "connect", "delete", "get", "head", "options", "patch", "post", "put", and "other".

I'm guessing that we should match that.

xeenon replied:
Yeah, that would be fine — 4 bits with room for more if HTTP adds more methods.

xeenon replied:
We talked about this offline, and we do need to support multiple request methods since the input is an array (set).

xeenon commented on Source/WebCore/loader/ResourceLoadInfo.cpp:
```diff
@@ -152,18 +152,42 @@ std::optional<OptionSet<LoadContext>> readLoadContext(StringView name)
     return std::nullopt;
 }
 
+std::optional<OptionSet<RequestMethod>> readRequestMethod(StringView name)
```
This would be `RequestMethod`, not `OptionSet` since only one method is used at a time.
```suggestion
std::optional<RequestMethod> readRequestMethod(StringView name)
```

elijahsawyers replied:
Is there any harm in leaving it as an `OptionSet`? It allows us to reuse the helper methods when parsing the JSON. E.g. `getTypeFlags(const JSON::Array& array, ResourceFlags& flags, std::optional<OptionSet<T>> (*stringToType)(StringView))`.

elijahsawyers commented on Source/WebCore/contentextensions/ContentExtensionsBackend.cpp:
```diff
@@ -340,7 +341,7 @@ ContentRuleListResults ContentExtensionsBackend::processContentRuleListsForLoad(
 ContentRuleListResults ContentExtensionsBackend::processContentRuleListsForPingLoad(const URL& url, const URL& mainDocumentURL, const URL& frameURL)
 {
-    ResourceLoadInfo resourceLoadInfo { url, mainDocumentURL, frameURL, ResourceType::Ping };
+    ResourceLoadInfo resourceLoadInfo { url, mainDocumentURL, frameURL, ResourceType::Ping, false, { } };
```
~This should use request method of post (per Google's AI summary). I need to double check.~

I don't think this is using HTTP at all. @achristensen07 Can you confirm what we should do here?

achristensen07 replied:
NetworkLoadChecker::processContentRuleListsForLoad has a ResourceRequest, which has a method.

elijahsawyers commented on Source/WebCore/contentextensions/ContentExtensionsBackend.cpp:
```diff
@@ -372,7 +373,7 @@ ContentRuleListResults ContentExtensionsBackend::processContentRuleListsForPingL
 bool ContentExtensionsBackend::processContentRuleListsForResourceMonitoring(const URL& url, const URL& mainDocumentURL, const URL& frameURL, OptionSet<ResourceType> resourceType)
 {
-    ResourceLoadInfo resourceLoadInfo { url, mainDocumentURL, frameURL, resourceType };
+    ResourceLoadInfo resourceLoadInfo { url, mainDocumentURL, frameURL, resourceType, false, { } };
```
@achristensen07 What should the request method be here? What is resource monitoring?

achristensen07 replied:
@basuke should know more than I do about this.

basuke replied:
We won't use request method so it can be anything. But isn't OptionSet's default constructor same with { }? If you can omit them, I think untouching is better than explicit change.

achristensen07 commented on Source/WebCore/loader/ResourceLoadInfo.h:
```diff
@@ -72,19 +72,33 @@ enum class LoadContext : uint32_t {
 
+enum class RequestMethod : uint32_t {
+    Get = 0x200000,
+    Head = 0x400000,
```
That's a lot of bits.  Do we need to support the specification of multiple methods, or would it suffice to say "only apply this rule to get requests"?

elijahsawyers replied:
We considered making the `request-method` take a string rather than an array of strings so that we could use 4 bits instead of 9. However, on the dNR side of things, this would require creating multiple WebKit rules if there are multiple request methods.

I think we can probably just trim down the request methods we explicitly support and put everything else in an "other" case if you feel strongly about this. Otherwise, I think we'd prefer to keep this correct by having all request methods, though. We can investigate increasing the size later, if needed.

achristensen07 replied:
I feel strongly about this.  We are running out of bits.  4 bits instead of 9 is well worth almost anything.

elijahsawyers replied:
Updated the patch to make the request methods mutually exclusive and only require 4 extra bits instead of 9.

achristensen07 commented on Source/WebCore/loader/ResourceLoadInfo.h:
```diff
@@ -72,19 +72,33 @@ enum class LoadContext : uint32_t {
+// The next 30 bits are used for the flags (ResourceType, LoadType, LoadContext, ActionCondition).
 // The values -1 and -2 are used for removed and empty values in HashTables.
-static constexpr uint64_t ActionFlagMask = 0x001FFFFF00000000;
+static constexpr uint64_t ActionFlagMask = 0x3FFFFFFF00000000;
```
Bits don't grow on trees.  If we can reduce the bit use that'd be ideal.

elijahsawyers replied:
https://github.com/WebKit/WebKit/pull/43503#discussion_r2038569345

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ContentRuleListNotification.mm:
```diff
@@ -348,6 +348,95 @@ HTTPServer beaconServer({
+        TestWebKitAPI::Util::spinRunLoop();
+    [webView stopLoading];
+
+    request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:@"apitest:///"]];
```
This is very repetitive.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ContentRuleListNotification.mm:
```diff
@@ -348,6 +348,95 @@ HTTPServer beaconServer({
+    [webView stopLoading];
+
+    Vector<Notification> expectedNotifications {
+        { "blockList"_s, "apitest:///"_s, true, false, false, { } },
```
These expected notifications indicate it doesn't test useful negative cases.

elijahsawyers replied:
I'll add the negative cases.

basuke commented on Source/WebCore/loader/ResourceMonitor.cpp:
```diff
@@ -112,7 +112,9 @@ void ResourceMonitor::didReceiveResponse(const URL& url, OptionSet<ContentExtens
-        .type = resourceType
+        .type = resourceType,
+        .mainFrameContext = false,
+        .requestMethod = { }
```
Ditto.

basuke reviewed (commented):
LGTM from the perspective of resource monitoring with a bit of change request. But I cannot approve it for entire Content Extension part logic.

achristensen07 commented on Source/WebCore/contentextensions/ContentExtensionsBackend.cpp:
```diff
@@ -235,6 +235,7 @@ ContentRuleListResults ContentExtensionsBackend::processContentRuleListsForLoad(
     URL mainDocumentURL;
     URL frameURL;
     bool mainFrameContext = false;
+    RequestMethod requestMethod = readRequestMethod(initiatingDocumentLoader.request().httpMethod().convertToASCIILowercase()).value();
```
.value() will crash if it's an unexpected value.  Let's use .value_or(RequestMethod::None) instead to be more robust against unexpected entries.

Also, convertToASCIILowercase does a little bit of unnecessary work.  What would you think about using equalIgnoringASCIICase in readRequestMethod instead?  It would make the parsing a little more flexible, but it's also pretty common to have all caps request methods.

achristensen07 commented on Source/WebCore/contentextensions/ContentExtensionsBackend.cpp:
```diff
@@ -338,9 +339,10 @@ ContentRuleListResults ContentExtensionsBackend::processContentRuleListsForLoad(
+ContentRuleListResults ContentExtensionsBackend::processContentRuleListsForPingLoad(const URL& url, const URL& mainDocumentURL, const URL& frameURL, const String& httpMethod)
 {
-    ResourceLoadInfo resourceLoadInfo { url, mainDocumentURL, frameURL, ResourceType::Ping };
+    RequestMethod requestMethod = readRequestMethod(httpMethod.convertToASCIILowercase()).value();
```
Same about value_or

elijahsawyers committed:
Implement request methods content blocker trigger field
https://bugs.webkit.org/show_bug.cgi?id=290962
rdar://148474901

Reviewed by Alex Christensen.

This patch introduces a new trigger field, request-method, to perform an action
only if the HTTP request method matches the specified HTTP request method. This
field has a string value, which can be any of the HTTP methods (lowercased) as
defined by:

https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Methods

We chose to use a string value, rather than an array of strings, because making
the request methods mutually exclusive saves more space (4 bits instead of 9).
Although this pattern deviates from the trigger field norm, it's better to save
the bits since we're running out of room.

The motivation behind this change is that we need to support requestMethods and
excludedRequestMethods in dNR, which isn't possible without this change.

* Source/WebCore/contentextensions/ContentExtensionError.cpp:
(WebCore::ContentExtensions::contentExtensionErrorCategory):
* Source/WebCore/contentextensions/ContentExtensionError.h:
Introduce a new error for invalid request-method field.

* Source/WebCore/contentextensions/ContentExtensionParser.cpp:
(WebCore::ContentExtensions::loadTrigger):
Parse the request-method into the trigger flags, raising an error as needed.

* Source/WebCore/contentextensions/ContentExtensionsBackend.cpp:
(WebCore::ContentExtensions::ContentExtensionsBackend::processContentRuleListsForLoad):
(WebCore::ContentExtensions::ContentExtensionsBackend::processContentRuleListsForPingLoad):
Parse the HTTP request method of the load and include it in the resource load
info.

* Source/WebCore/contentextensions/ContentExtensionsBackend.h:
Make the method for processing content rule lists for ping loads take in an HTTP
method parameter.

* Source/WebCore/contentextensions/DFABytecode.h:
* Source/WebCore/contentextensions/DFABytecodeCompiler.cpp:
(WebCore::ContentExtensions::bytecodeFlagsSize):
(WebCore::ContentExtensions::toSizeT):
* Source/WebCore/contentextensions/DFABytecodeInterpreter.cpp:
(WebCore::ContentExtensions::consumeResourceFlagsAndInstruction):
(WebCore::ContentExtensions::DFABytecodeInterpreter::interpretTestFlagsAndAppendAction):
Introduce a 32 bit flags size. With the introduction of request methods, the
flags can be up to 25 bits in size.

* Source/WebCore/loader/ResourceLoadInfo.cpp:
(WebCore::ContentExtensions::readRequestMethod):
(WebCore::ContentExtensions::ResourceLoadInfo::getResourceFlags const):
* Source/WebCore/loader/ResourceLoadInfo.h:
(WebCore::ContentExtensions::ResourceLoadInfo::isolatedCopy const):
(WebCore::ContentExtensions::ResourceLoadInfo::isolatedCopy):
Introduce the request methods enum, and modify the byte code as necessary.

* Source/WebKit/NetworkProcess/NetworkLoadChecker.cpp:
(WebKit::NetworkLoadChecker::processContentRuleListsForLoad):
Pass in the HTTP method when processing content rule lists for ping load.

* Tools/TestWebKitAPI/Tests/WebCore/ContentExtensions.cpp:
(TestWebKitAPI::TEST_F(ContentExtensionTest, RequestMethod)):
Write a new test for validating the behavior of the request-method field.

(TestWebKitAPI::TEST_F(ContentExtensionTest, InvalidJSON)):
Add assertions for validating the request-method field in the JSON.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/ContentRuleListNotification.mm:
(TEST(ContentRuleList, RequestMethods)):
Write a new test for validating the behavior of the request-method field.

Canonical link: https://commits.webkit.org/293746@main
